src/types/timestamp.rs

```diff
@@ -1007,230 +973,212 @@ impl TimestampBuilder {
 
     // 0-indexed day
-    pub fn with_day0(self, day: u32) -> HourAndMinuteSetter {
+    pub fn with_day0(self, day: u32) -> TimestampBuilder<HasDay> {
         self.with_day(day + 1)
     }
 
     // 1-indexed day
-    pub fn with_day(self, day: u32) -> HourAndMinuteSetter {
-        let mut builder = self.builder;
-        builder.precision = Precision::Day;
-        builder.day = Some(day as u8);
-        HourAndMinuteSetter { builder }
-    }
-
-    pub fn build(self) -> IonResult<Timestamp> {
-        self.into_builder().build()
+    pub fn with_day(mut self, day: u32) -> TimestampBuilder<HasDay> {
+        self.precision = Precision::Day;
+        self.day = Some(day);
+        self.change_state()
     }
 }
 
-/// Allows the user to set the `hour` and `minute` fields on a builder that has already
-/// had its `year`, `month`, and `day` fields set. Or, if `Day` is the desired precision,
-/// they may build the [Timestamp] with an unknown offset instead.
 #[derive(Debug, Clone)]
-pub struct HourAndMinuteSetter {
-    builder: TimestampBuilder,
-}
-
-impl HourAndMinuteSetter {
-    pub fn with_hms(self, hour: u32, minute: u32, second: u32) -> FractionalSecondSetter {
-        let mut builder = self.builder;
-        builder.hour = Some(hour as u8);
-        builder.minute = Some(minute as u8);
-        builder.second = Some(second as u8);
-        builder.precision = Precision::Second;
-        FractionalSecondSetter { builder }
+pub struct HasDay;
+impl TimestampBuilder<HasDay> {
+    pub fn with_hms(self, hour: u32, minute: u32, second: u32) -> TimestampBuilder<HasSeconds> {
+        self.with_hour(hour).with_minute(minute).with_second(second)
     }
 
-    pub fn with_hour_and_minute(self, hour: u32, minute: u32) -> SecondSetter {
-        let mut builder = self.builder;
-        builder.precision = Precision::HourAndMinute;
-        builder.hour = Some(hour as u8);
-        builder.minute = Some(minute as u8);
-        SecondSetter { builder }
+    pub fn with_hour_and_minute(mut self, hour: u32, minute: u32) -> TimestampBuilder<HasMinute> {
+        self.precision = Precision::HourAndMinute;
+        self.hour = Some(hour);
+        self.minute = Some(minute);
+        self.change_state()
     }
 
-    pub fn build(self) -> IonResult<Timestamp> {
-        self.into_builder().build()
+    pub fn with_hour(mut self, hour: u32) -> TimestampBuilder<HasHour> {
+        self.precision = Precision::HourAndMinute;
+        self.hour = Some(hour);
+        self.minute = Some(0);
+        self.change_state()
     }
 }
 
-/// Allows the user to set the `second` field on a builder that has already
-/// had its `year`, `month`, `day`, `hour`, and `minute` fields set. Or, if `HourAndMinute` is the
-/// desired precision, they may build the [Timestamp] instead, optionally specifying an offset if
-/// known.
-#[derive(Debug, Clone)]
-pub struct SecondSetter {
-    builder: TimestampBuilder,
+macro_rules! with_offset {
```

Comment:
🗺️ We _could_ add this function to particular states by using a trait impl, but I think it makes the rust doc page more readable to just have a macro to "paste" it into the appropriate impl blocks.